x-pack/plugins/fleet/server/services/agents/crud.ts

```diff
@@ -122,30 +122,48 @@ export async function getAgentsByKuery(
 
   const kueryNode = _joinFilters(filters);
   const body = kueryNode ? { query: toElasticsearchQuery(kueryNode) } : {};
-  const res = await esClient.search<FleetServerAgent, {}>({
-    index: AGENTS_INDEX,
-    from: (page - 1) * perPage,
-    size: perPage,
-    track_total_hits: true,
-    ignore_unavailable: true,
-    body: {
-      ...body,
-      sort: [{ [sortField]: { order: sortOrder } }],
-    },
-  });
+  const queryAgents = async (from: number, size: number) =>
+    esClient.search<FleetServerAgent, {}>({
+      index: AGENTS_INDEX,
+      from,
+      size,
+      track_total_hits: true,
+      ignore_unavailable: true,
+      body: {
+        ...body,
+        sort: [{ [sortField]: { order: sortOrder } }],
+      },
+    });
+  const res = await queryAgents((page - 1) * perPage, perPage);
 
   let agents = res.body.hits.hits.map(searchHitToAgent);
+  let total = (res.body.hits.total as estypes.SearchTotalHits).value;
   // filtering for a range on the version string will not work,
   // nor does filtering on a flattened field (local_metadata), so filter here
   if (showUpgradeable) {
-    agents = agents.filter((agent) =>
-      isAgentUpgradeable(agent, appContextService.getKibanaVersion())
-    );
+    // fixing a bug where upgradeable filter was not returning right results https://github.com/elastic/kibana/issues/117329
+    // query all agents, then filter upgradeable, and return the requested page and correct total
+    // if there are more than MAX_SIZE agents, the logic falls back to same as before
+    const MAX_SIZE = 5000; // choosing a high size to cover most scenarios
+    if (total < MAX_SIZE) {
+      const response = await queryAgents(0, MAX_SIZE);
+      agents = response.body.hits.hits.map(searchHitToAgent);
+      agents = agents.filter((agent) =>
```

Comment:
It's a little odd that we reassign `agents` twice in a row here. We could instead chain the `.filter` onto the `.map(searchHitToAgent)` and do away with the intermediate assignment.

Reply:
thanks for the review, fixed both comments